flox-build.mk: "pure" builds fail when deleting tracked files

**Describe the bug:**

`git ls-files` will list all _tracked_ files **including deleted files**.
Consequently, when we try to create a tarball with all files listed by `git ls-files` we may attempt packing files that are actually deleted:


**Steps to reproduce:**
1. `git init` && `flox init`
2. git touch file && git add file && rm file
3. add any trivial build
4. `flox build`

Result:

```
$ flox build foo
Rendering foo build script to /tmp/nix-shell.KfqmHe/712176e7/foo/build.bash
/nix/store/lpxsvgscb0i36hfp3q6m87ppcnin966i-gnutar-1.35/bin/tar: file: Cannot stat: No such file or directory
/nix/store/lpxsvgscb0i36hfp3q6m87ppcnin966i-gnutar-1.35/bin/tar: Exiting with failure status due to previous errors
make: *** [...]/cli/build/flox-package-builder/libexec/flox-build.mk:717: /tmp/nix-shell.KfqmHe/712176e7/foo/src.tar] Error 2
❌ ERROR: Build failed
```

Expected:

Successful build


**Flox Version (run `flox --version` if possible):**

1.6.0-g6a736f171

--- a/cli/flox-rust-sdk/src/providers/build.rs
+++ b/cli/flox-rust-sdk/src/providers/build.rs
@@ -3201,14 +3201,7 @@ mod tests {
         git.add(&[&deleted_file_path]).unwrap();
         fs::remove_file(&deleted_file_path).unwrap();
 
-        let output = assert_build_status(&flox, &mut env, package_name, None, false);
-        assert!(
-            output.stderr.contains(&format!(
-                "/bin/tar: {deleted_file_name}: Cannot stat: No such file or directory"
-            )),
-            "stderr: {}",
-            output.stderr
-        )
+        assert_build_status(&flox, &mut env, package_name, None, true);
     }
 }
 

--- a/package-builder/flox-build.mk
+++ b/package-builder/flox-build.mk
@@ -51,6 +51,7 @@ __package_bin = $(if $(filter @%@,$(1)),$(2),$(1)/bin/$(2))
 _bash := $(call __package_bin,$(__bashInteractive),bash)
 _cat := $(call __package_bin,$(__coreutils),cat)
 _chmod := $(call __package_bin,$(__coreutils),chmod)
+_comm := $(call __package_bin,$(__coreutils),comm)
 _cp := $(call __package_bin,$(__coreutils),cp)
 _cpio := $(call __package_bin,$(__cpio),cpio)
 _cut := $(call __package_bin,$(__coreutils),cut)
@@ -490,15 +491,24 @@ define BUILD_nix_sandbox_template =
   # build cache.  This is used for (at least) publish.
   $(eval _do_buildCache = $(if $(DISABLE_BUILDCACHE),,true))
 
+  # 'git ls-files' will list all _tracked_ files **including deleted files**.
+  # Consequently, when we try to create a tarball with all files listed by
+  # 'git ls-files' we may attempt packing files that are actually deleted.
+  # To avoid this, we filter out deleted files.
+  # Because 'git ls-files' does not have a flag to filter deleted files,
+  # but allows to _only_ show deleted files, use 'comm' to do the filtering for
+  # us.
+  # Assumes the output of 'git ls-files' to be sorted.
+  $(eval $(_pvarname)_src_list = $($(_pvarname)_tmpBasename)/src-list)
+  $($(_pvarname)_src_list): $(PROJECT_TMPDIR)/check-build-prerequisites
+	$(_comm) -23 <($(_git) ls-files -c) <($(_git) ls-files -d) > $$@
+
   # The sourceTarball value needs to be stable when nothing changes across
   # builds, so we create a tarball at a stable temporary path and pass that
   # to the derivation instead.
   $(eval $(_pvarname)_src_tar = $($(_pvarname)_tmpBasename)/src.tar)
-  $($(_pvarname)_src_tar): $(PROJECT_TMPDIR)/check-build-prerequisites
-	@# TIL that you have to explicitly call `wait` to harvest the exit status
-	@# of a process substitution, and that `set -o pipefail` does nothing here.
-	@# See: https://mywiki.wooledge.org/ProcessSubstitution
-	$(_V_) $(_tar) -cf $$@ --no-recursion -T <($(_git) ls-files) && wait "$$$$!"
+  $($(_pvarname)_src_tar): $($(_pvarname)_src_list)
+	$(_V_) $(_tar) -cf $$@ --no-recursion --files-from $$<
 
   # The buildCache value needs to be similarly stable when nothing changes across
   $(eval $(_pvarname)_buildCache = $($(_pvarname)_tmpBasename)/buildCache.tar)

